sockeye/training.py

```diff
@@ -275,16 +299,31 @@ def _fit(self,
                         os.remove(best_path)
                     actual_best_fname = C.PARAMS_NAME % best_checkpoint
                     os.symlink(actual_best_fname, best_path)
-                    num_not_improved = 0
+                    training_state.num_not_improved = 0
                 else:
-                    num_not_improved += 1
-
-                if num_not_improved == max_num_not_improved:
-                    logger.info("Model has not improved for %d checkpoints. Stopping fit.", num_not_improved)
-                    training_monitor.stop_fit_callback()
+                    training_state.num_not_improved += 1
+
+                if training_state.num_not_improved == max_num_not_improved:
+                    logger.info("Model has not improved for %d checkpoints. Stopping fit.", training_state.num_not_improved)
+                    self.training_monitor.stop_fit_callback()
+                    final_training_state_dirname = os.path.join(output_folder, C.TRAINING_STATE_DIRNAME)
+                    if os.path.exists(final_training_state_dirname):
+                        shutil.rmtree(final_training_state_dirname)
                     break
+                else:
+                    self._checkpoint(training_state, output_folder, train_iter)
 
-    def _evaluate(self, checkpoint, val_iter, val_metric, training_monitor):
+    def _save_params(self, output_folder: str, checkpoint: int):
+        """
+        sync aux params across devices
```

Comment:
This method doesn't really sync aux params across devices, the module.get_params() call does.

Since this is an internal method we dont need to have a docstring necessarily, but if we have one, it should describe what the method overall does (saving the current module parameters to a file)